src/services/download/mod.rs

```diff
@@ -0,0 +1,173 @@
+            }
+        });
+        let fut01 = fut03
+            .map(|spawn_ret| spawn_ret.unwrap_or_else(|_| Err(DownloadErrorKind::Canceled.into())))
+            .boxed()
+            .compat();
+        Box::new(future_metrics!(
+            "service.download",
+            Some((
+                std::time::Duration::from_secs(3600),
+                DownloadErrorKind::Canceled.into()
+            )),
+            fut01
+        ))
+    }
+}
+
+/// Download the source from a streaming future.
+///
+/// These streaming futures are currently implemented per source type.
+fn download_future_stream(
+    source_desc: String,
+    stream: Box<dyn Future<Item = Option<DownloadStream>, Error = DownloadError>>,
+    destination: PathBuf,
+) -> Box<dyn Future<Item = DownloadStatus, Error = DownloadError>> {
+    // All file I/O in this function is blocking!
+    let ret = stream.and_then(move |maybe_stream| match maybe_stream {
+        Some(stream) => {
+            log::trace!("Downloading from {}", source_desc);
+            let file = tryf!(
+                std::fs::File::create(&destination).context(DownloadErrorKind::BadDestination)
+            );
+            let fut = stream
+                .fold(file, |mut file, chunk| {
+                    file.write_all(chunk.as_ref())
+                        .context(DownloadErrorKind::Write)
+                        .map_err(DownloadError::from)
+                        .map(|_| file)
+                })
+                .and_then(|_| Ok(DownloadStatus::Completed));
+            Box::new(fut) as Box<dyn Future<Item = DownloadStatus, Error = DownloadError>>
+        }
+        None => {
+            let fut = future::ok(DownloadStatus::NotFound);
+            Box::new(fut) as Box<dyn Future<Item = DownloadStatus, Error = DownloadError>>
+        }
+    });
+    Box::new(ret)
+}
+
+#[cfg(test)]
+mod tests {
+    // Actual implementation is tested in the sub-modules, this only needs to
+    // ensure the service interface works correctly.
+    use super::*;
+
+    use crate::sources::SourceConfig;
+    use crate::test;
+
+    #[test]
+    fn test_download() {
+        test::setup();
+
+        // test::setup() enables logging, but this test spawns a thread where
+        // logging is not captured.  For normal test runs we don't want to
+        // pollute the stdout so silence logs here.  When debugging this test
+        // you may want to temporarily remove this.
+        log::set_max_level(log::LevelFilter::Off);
+
+        let tmpfile = tempfile::NamedTempFile::new().unwrap();
+        let dest = tmpfile.path().to_owned();
+
+        let (_srv, source) = test::symbol_server();
+        let source_id = match source {
+            SourceConfig::Http(source) => {
+                SourceFileId::Http(source, SourceLocation::new("hello.txt"))
+            }
+            _ => panic!("unexpected source"),
+        };
+
+        let dl_svc = DownloadService::new(RemoteThread::new_threaded());
```

Comment:
We can't get rid of `::new_threaded()` here because `crate::utils::futures::IS_TEST` is global and could affect other concurrent tests.  So I think `::new_threaded()` is fine.